enh: added possibility to compute spin-degenerate calculations

sofiasanz opened the pull request:
With these changes now one should be able to run spin-degenerate calculations. Basically now the dimension of the problem is determined by the dimension of the tight-binding `sisl.Hamiltonian`. I modified `test-hubbard-quick.py` to test the functions for a spin-degenerate calculation of a molecule.

tfrederiksen commented on hubbard/hamiltonian.py:
```diff
@@ -259,13 +264,13 @@ def update_hamiltonian(self):
         q0 = self.geometry.atoms.q0
         E = self.e0.copy()
-        E += self.U * (self.n[[1, 0], :] - q0)
+        E += self.U * (self.n[::-1, :] - q0)
```
This may be problematic if the spin dimensionality is higher than two and/or if the basis is non-orthogonal (where the last element is the overlap).

tfrederiksen reviewed (changes requested):
Very nice Sofia! This branch addresses #33 .

My only general comment is that the loops of type `for s in range(H.H.dim)` will fail if the user provides a TB hamiltonian of a higher spin class than 2 and/or if the basis is set to non-orthogonal (then `dim = nspins + 1`, the last index being the overlap). I am wondering if `hubbard` should be able to handle such situations too?

You can access the spin degrees of freedom in the TB hamiltonian via `H.spin.spins`, this will avoid the possible extra dimension if the overlaps are explicitly given.

tfrederiksen commented on hubbard/hamiltonian.py:
```diff
@@ -24,13 +24,15 @@ class HubbardHamiltonian(object):
-        A spin-polarized tight-binding Hamiltonian
+        A tight-binding Hamiltonian
     n: numpy.ndarray, optional
         initial spin-densities vectors. The shape of `n` must be (2, len(geometry))
```
also true if `spin=unpolarized`?

tfrederiksen commented on hubbard/hamiltonian.py:
```diff
@@ -62,13 +61,18 @@ def __init__(self, TBHam, n=0, U=0.0, q=(0., 0.), nkpt=[1, 1, 1], kT=1e-5):
         self.q = np.array(q, dtype=np.float64).copy()
-        assert len(self.q) == 2 # Users *must* specify two values
+        # Ensure the length of q is equal to the dimension of the problem
+        self.q = self.q[:self.H.dim]
```
`self.q = self.q[:self.H.spin.spins]`?

zerothi replied:
Actually it needs to be `min(2, self.H.spin.spins)`.

I will add a new flag to `Spin` to get the number of spinors

Then it would be `self.q = self.q[:self.H.spin.spinor]`

sofiasanz replied:
Now this is handled by saving the spin degree of freedom in the `HubbardHamiltonian` object, and by setting a maximum of 2 since we have only implemented the cases for either `polarized` or `unpolarized` Hamiltonians.
See commit a5632f0.

tfrederiksen commented:
Another thought would be to let the user provide *any* spin-type for the TB hamiltonian, and then let a keyword `HubbardHamiltonian(..., spin={'polarized', 'unpolarized'})` define how the calculation proceeds? Then internally in the initialization phase, we can use the new `transform` function in sisl as follows:
```python
self.H = TBHam.transform(spin=spin)
```
and we can then loop over `for s in self.H.spin.spins`.

zerothi commented:
I agree on Thomas' view. Generally one should not use the `H.dim` property since it depends on orthogonality etc.
Using `spin.*` is much better.

But, I don't think that the `transform` should be done *inside* `HubbardHamiltonian`, it just seems weird to me.
If a users wants a polarized calculation, then use `sisl.Hamiltonian(spin='pol')` if an unpolarized, use `sisl.Hamiltonian()`.

Also, there should probably be tests against spin-orbit/non-collinear. They still don't work as far as I can see. However, what you probably then want is some vector describing the polarization of the spin?

If you want to go down that route possibly it would be better have an interface looking like this (just a thought!):
```
HubbardHamiltonian(q=<total # of electrons>, pol=[0, 0, 1])
```
where the length of the polarization determines the polarization. Since the `Geometry` carries the total number of electrons, you might not even need `q` argument.

sofiasanz commented:
Hi Thomas and Nick, thank you very much for the comments! I was not aware that the dimension of the Hamiltonian also holds the orthogonality. I will change that and use `spins` instead.

I agree that the Hamiltonian should be determined as spin-polarized or non-polarized when building the tight-binding Hamiltonian rather than in the `HubbardHamiltonian` class.

> Also, there should probably be tests against spin-orbit/non-collinear. They still don't work as far as I can see. However, what you probably then want is some vector describing the polarization of the spin?

Yeah we don't have a method to obtain the spin densities for a non-collinear/spin-orbit, but I guess if we know how to build the spin densities for these Hamiltonians the SCF process should be the same, no?


> If you want to go down that route possibly it would be better have an interface looking like this (just a thought!):
> 
> ```
> HubbardHamiltonian(q=<total # of electrons>, pol=[0, 0, 1])
> ```

I'm not sure if I follow this idea! What are the three indices of the polarization?

> where the length of the polarization determines the polarization. Since the Geometry carries the total number of electrons, you might not even need q argument.

Well the thing is that then we should ensure that the user set the correct charge when building the geometry, because otherwise it may not give the correct solution. That's why we have the definition of the total charge by default. And also to define a possible spin imbalance.

zerothi commented:
The point of `pol` is that you could *force* a polarization along a given spin-direction. In the polarized case it would always be along z direction.
For nc and soc this ain't trivial because it requires the bands around the Fermi level to have *many* different angles to be able to get it. So probably not worth it for now. Just a thought...

zerothi commented:
As for building DM for nc and soc, this ain't too difficult. :)

sofiasanz commented:
I see! Yeah maybe for the future, but is nice to think about it :)

> As for building DM for nc and soc, this ain't too difficult. :)

Could it be done like computing the expected value of the S_z operator for the Hamiltonian or something like this I guess?

zerothi commented:
> > As for building DM for nc and soc, this ain't too difficult. :)
> 
> Could it be done like computing the expected value of the S_z operator for the Hamiltonian or something like this I guess?

Yes. That would be a good initial implementation I think. However, it is both the diagonal part of the DM + the spin moment.
So you'll need both S_x, S_y and S_z to get the correct spin-configuration.

sofiasanz commented:
I see. I should read more about it! But it could be pretty cool if we'd have something for these systems too.

sofiasanz commented on hubbard/plot/charge.py:
```diff
@@ -140,6 +140,9 @@ def __init__(self, HH, ext_geom=None, realspace=False, **kwargs):
+            raise ValueError('To plot the spin polarization the Hamiltonian needs to be spin-polarized!')
+
         # Compute charge difference between up and down channels
         chg = np.diff(HH.n[[1, 0]], axis=0).ravel()
```
Now this function plots the spin-polarization for both `polarized` and `unpolarized` Hamiltonians.

sofiasanz commented on hubbard/hamiltonian.py:
```diff
@@ -62,13 +61,18 @@ def __init__(self, TBHam, n=0, U=0.0, q=(0., 0.), nkpt=[1, 1, 1], kT=1e-5):
+                self.q[1] = ntot / 2
+            if self.q[0] <= 0:
+                self.q[0] = ntot - self.q[1]
 
```
I changed these lines a little bit but I wanted to keep what we had before: for the polarized Hamiltonian, if there is an odd number of electrons set the largest number to up electrons (`q[0]`).

sofiasanz commented on hubbard/hamiltonian.py:
```diff
@@ -24,13 +24,15 @@ class HubbardHamiltonian(object):
         on-site Coulomb repulsion
     q: array_like, optional
-        Two values specifying up, down electron occupations
+        One or two values specifiyng the total charge associated to each spin component.
```
Thanks!

sofiasanz commented:
@tfrederiksen  @zerothi Thank you very much for your comments! Now I use the spin degree of freedom (which is upper-bounded by 2) to determine the dimension of the relevant quantities (spin-densities, charges, etc.).
All these changes are collected in commit a5632f0.
What do you think?

zerothi commented:
> @tfrederiksen @zerothi Thank you very much for your comments! Now I use the spin degree of freedom (which is upper-bounded by 2) to determine the dimension of the relevant quantities (spin-densities, charges, etc.).
> All these changes are collected in commit [a5632f0](https://github.com/dipc-cc/hubbard/commit/a5632f043f50fb0c0e73327bb71a71d482b13eb2).
> What do you think?

Great with all the work... But :)  
`H.dim==3` for a polarized non-orthogonal Hamiltonian.  
`H.dim==2` for an unpolarized non-orthogonal Hamiltonian.

I think you should use `H.spin.spins == 2` and separately check for orthogonality, as @tfrederiksen mentions.

sofiasanz commented:
Hi Nick, in my latest commit now the spin degree of freedom is determined by `H.spin.spins` and not `H.dim` anymore. I checked and it works fine with non-orthogonal basis also.

tfrederiksen commented on tests/test-hubbard-quick.py:
```diff
@@ -23,10 +24,26 @@
 H.write_initspin('test.fdf', mode='w')
+
+import random
+print('5. Run one iteration for spin-degenerate calculattion')
```
typo

tfrederiksen commented on tests/test-hubbard-quick.py:
```diff
@@ -23,10 +24,26 @@
+dn = H.iterate(density.calc_n)
+print('   dn, Etot: ', dn, H.Etot, '\n')
+
+print('6. Run one iteration for spin-degenerate calculattion with NEGF')
```
same typo here

tfrederiksen commented on hubbard/sp2.py:
```diff
@@ -20,11 +20,39 @@ def sp2(ext_geom, t1=2.7, t2=0.2, t3=0.18, eB=3., eN=-3.,
+    eB: float, optional
+        on-site energy for Boron atoms
+    eN: float, optional
+        on-site energy for Nytrogen atoms
```
typo

tfrederiksen commented on hubbard/hamiltonian.py:
```diff
@@ -323,6 +331,8 @@ def fermi_level(self, q=[None, None], dist='fermi_dirac'):
         ----------
         q: array_like, optional
             charge per spin channel. First index for spin up, second index for dn
+            If the Hamiltonian is non-polarized q should have only one component
```
unpolarized?

tfrederiksen reviewed (approved):
Looks good Sofia! Just made some minor remarks. What about also adding a dedicated test that checks (for a system that does not polarize) that the self-consistent solutions lead to the same total energy for both `unpolarized` and `polarized` types of runs?

tfrederiksen commented on hubbard/hamiltonian.py:
```diff
@@ -55,20 +54,25 @@ def __init__(self, TBHam, n=0, U=0.0, q=(0., 0.), nkpt=[1, 1, 1], kT=1e-5):
         self.H.finalize()
         self.geometry = TBHam.geometry
+        # So far we only consider either unpolarized or spin-polarized Hamiltonians
+        self.spins = min(2, self.H.spin.spins)
```
See https://github.com/zerothi/sisl/commit/8ee0de851b52c4c117501045c6ccfd1b46fef6db.

zerothi replied:
Yeah, regarding `spin.spins` probably that is bad wording. It should more likely be `spin.components/elements`, or?

tfrederiksen replied:
> Yeah, regarding `spin.spins` probably that is bad wording. It should more likely be `spin.components/elements`, or?

`spin.spins` comes from sisl. Would you consider to change that?

zerothi replied:
Yes and no. I will retain it for backwards compatibility, but I think perhaps `Spin.size` is more appropriate. I have amended sisl to use `Spin.size` instead. If you have another suggestion, please let me know!

The idea is that it allows one to determine the number of spin-elements required to describe the spin-kind. I.e. a polarized requires 2 elements. Whereas non-collinear requires 4 reals, or 3 complex arrays. Here `Spin.spins` doesn't really make sense. So the current discussion led me to believe its name suggested another use-case.

sofiasanz replied:
I'm not sure which is the best name for our variable here. Maybe `self.spinors` can be a bit confusing? Since we are not using directly spinors here, and actually is the total number of spin components what we need. On the other hand, I'm using `self.spins = min(2, self.H.spin.spins)` because we only allow for spin polarized or unpolarized Hamiltonians, but if we had the implementation for SOC or NC then what would the dimension be?

tfrederiksen replied:
Maybe then follow sisl nomenclature and define `self.size = min(2, self.H.spin.size)`?

sofiasanz replied:
Yes I think that's the best :)

zerothi replied:
Yes, the `spin.spinor` is the number of diagonal elements in your spin box, that is either 1 (non polarized) or 2 (polarized, NC and SOC).

However, `spin.spins` doesn't really mean "number of spin components". It rather means, "number of numerical values needed to represent the spin-box". Therefore I think that `spin.spins` is not the best wording. So if you variable is used to signify spinors (diagonal components in the spin-box), then simply using `spin.spinor` would suffice.

sofiasanz replied:
Yes I agree. So I was thinking on renaming the variable to `self.size` following `sisl` notation, but this makes sense if the dimension of the problem (e.g. the dimension of the spin-densities) then is determined by the  "number of numerical values needed to represent the spin-box" rather than the number of "spin components", which is what I am not entirely sure (in that case I guess it would make more sense to use `self.spinors` as you both point out)...

zerothi replied:
Yeah, with @tfrederiksen suggestion you might as well use `Spin.spinor` since that is the same. I don't know exactly how +U looks for NC/SOC.

zerothi commented:
> Hi Nick, in my latest commit now the spin degree of freedom is determined by `H.spin.spins` and not `H.dim` anymore. I checked and it works fine with non-orthogonal basis also.

Sorry about my obviously wrong comment then! The email only showed the original changes, I didn't get that... Sorry!

sofiasanz commented on hubbard/hamiltonian.py:
```diff
@@ -496,7 +509,7 @@ def iterate(self, calc_n_method, q=None, mixer=None, **kwargs):
 
         # Store total energy
-        self.Etot = Etot - self.U * np.multiply.reduce(self.n, axis=0).sum()
+        self.Etot = Etot - self.U * (ni[0]*ni[-1]).sum()
```
Because if the dimension of this array is just one then it would crash. Of course this is only valid for polarized or unpolarized cases.

tfrederiksen replied:
OK, good point!

sofiasanz commented:
In commit d1c9d90 I changed the name of the variable `self.spins` to `self.size` and now it is equal to `H.spin.spinor`. I also added a test file (in commit 77be21d ) where we compute the reference molecule with `U=2` eV (for which the ground state is a closed-shell solution), so we can compare the solution obtained with a polarized or unpolarized Hamiltonian.

sofiasanz commented:
About the name of the variable `self.size` that comes from `H.spin.spinor` we can modify it afterwards when we think about how we can add the mean-field Hubbard term to a Hamiltonian with a SOC or NC term.

sofiasanz commented:
If you agree I will proceed on merging this branch :)

tfrederiksen commented on hubbard/hamiltonian.py:
```diff
@@ -55,20 +54,25 @@ def __init__(self, TBHam, n=0, U=0.0, q=(0., 0.), nkpt=[1, 1, 1], kT=1e-5):
         self.H.finalize()
         self.geometry = TBHam.geometry
+        # So far we only consider either unpolarized or spin-polarized Hamiltonians
+        self.size = self.H.spin.spinor
```
Without further labelling I find `size` not so descriptive. Maybe `self.spinor` or `self.spin_size` would be clearer?

sofiasanz replied:
Yeah, maybe is better to call it `self.spin_size`. To me `spinor` sounds like it could be confusing. I will change then the variable name.

tfrederiksen commented on hubbard/hamiltonian.py:
```diff
@@ -692,6 +707,10 @@ def spin_contamination(self, ret_exact=False):
         """
+
+        if self.size<2:
+            raise ValueError('To obtain the spin contamination the Hamiltonian needs to be spin-polarized!')
```
Is this necessary? If the user insists on doing this, the answer is just zero spin contamination, no?

sofiasanz replied:
Well I thought that it didn't make much sense to obtain the spin contamination for the unpolarized case, but I can modify this function to work also for this case and return the zero spin contamination.

tfrederiksen replied:
I think that would be better than raising errors. I see it more as a trivial calculation (rather than an error).

sofiasanz replied:
Done! See commit 75d58b9
